SCClassLibrary/Common/Audio/SynthDef.sc

```diff
@@ -1,105 +1,422 @@
+	classvar <synthDefDir;
+	classvar <>warnAboutLargeSynthDefs = false;
+
+	classvar <optimizationLevel;
+
+	var <>name;
+	var <>func;
 
 	var <>controls, <>controlNames; // ugens add to this
 	var <>allControlNames;
 	var <>controlIndex=0;
+
+	// This is the final output in topological order.
+	// It is also the initial order specified by adding UGens through evaluating the 'func'.
 	var <>children;
 
+	// Numbers.
 	var <>constants;
 	var <>constantSet;
 	var <>maxLocalBufs;
-
-	// topo sort
-	var <>available;
 	var <>variants;
-	var <>widthFirstUGens;
+
+	// This is nil in normal use, but true when optimisations are being ran.
 	var rewriteInProgress;
 
-	var <>desc, <>metadata;
+	var <>desc;
+	var <>metadata;
 
-	classvar <synthDefDir;
-	classvar <>warnAboutLargeSynthDefs = false;
+	// These are the ends of the graph, the Outs, BufWrs, et cetera.
+	var effectiveUGens;
+	// This stores all resource managers in an Event<Class, ImplicitResourceConnectionStrategy>.
+	var resourceManagers;
 
-	*synthDefDir_ { arg dir;
-		if (dir.last.isPathSeparator.not )
-			{ dir = dir ++ thisProcess.platform.pathSeparator };
+	*synthDefDir_ { |dir|
+		if (dir.last.isPathSeparator.not ) { dir = dir ++ thisProcess.platform.pathSeparator };
 		synthDefDir = dir;
 	}
 
+	*setOptimizations { |opts|
+		if (opts.isKindOf(SynthDefOptimizationFlags).not) { Error("opts should be a SynthDefOptimizationFlags").throw };
+		optimizationLevel = opts;
+	}
+
 	*initClass {
 		synthDefDir = Platform.userAppSupportDir ++ "/synthdefs/";
-		// Ensure exists:
 		synthDefDir.mkdir;
+		optimizationLevel = SynthDefOptimizationFlags.onlySorting; // default is just to sort.
+	}
+
+	*newForSynthDesc {
+		^super.newCopyArgs(
+			effectiveUGens: [],
+			resourceManagers: ImplicitResourceConnectionStrategy.createNewInstances,
+		)
 	}
 
 	*new { |name, ugenGraphFunc, rates, prependArgs, variants, metadata|
-	    ^super.newCopyArgs(
-	        name: name.asSymbol,
-	        variants: variants,
-	        metadata: metadata ?? {()},
-	        children: Array(64)
-        ).build(ugenGraphFunc, rates, prependArgs)
+		^super.newCopyArgs(
+			name: name.asSymbol,
+			func: ugenGraphFunc,
+			variants: variants,
+			metadata: metadata ?? {()},
+			children: Array(64),
+			effectiveUGens: [],
+			resourceManagers: ImplicitResourceConnectionStrategy.createNewInstances,
+		).build(rates, prependArgs)
+	}
+
+
+	// Same as new, but optimises for server-side performance. May take a long time to compile when the graph is large.
+	*newOptimize { |...args, kwargs|
+		^SynthDef.performArgs(\newWithOptimizations, [SynthDefOptimizationFlags.all] ++ args, kwargs)
+	}
+
+	*newWithOptimizations { |newOptimizations, name, ugenGraphFunc, rates, prependArgs, variants, metadata|
+		var out;
+		var oldOptimizations = optimizationLevel;
+		if (newOptimizations.isKindOf(SynthDefOptimizationFlags).not) { Error("newOptimizations should be a SynthDefOptimizationFlags").throw };
+		SynthDef.setOptimizations(newOptimizations);
```

Comment:
BTW -- almost overlooked this -- the pattern "1. Set something; 2. Do stuff; 3. Set back to the old thing" should pretty much always be wrapped in a protect block:

```
		protect {
			SynthDef.setOptimizations(newOptimizations);
			
			out = super.newCopyArgs(
				name: name,
				func: ugenGraphFunc,
				variants: variants,
				metadata: metadata ?? { () },
				children: Array(64),
				effectiveUGens: [],
				resourceManagers: ImplicitResourceConnectionStrategy.createNewInstances,
			).build(rates, prependArgs);
		} {
			SynthDef.setOptimizations(oldOptimizations);
		};
```

Otherwise, if SynthDef building errors out (which we can't be sure won't happen, since it's running a user function), then the optimization flags will have been overwritten.